include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,261 @@
+///       "triton cache API version does not support this cache");
+///   }
+///
+#define TRITONCACHE_API_VERSION_MAJOR 0
+#define TRITONCACHE_API_VERSION_MINOR 1
+
+/// Get the TRITONCACHE API version supported by Triton. This
+/// value can be compared against the
+/// TRITONCACHE_API_VERSION_MAJOR and
+/// TRITONCACHE_API_VERSION_MINOR used to build the cache to
+/// ensure that Triton is compatible with the cache.
+///
+/// \param major Returns the TRITONCACHE API major version supported
+/// by Triton.
+/// \param minor Returns the TRITONCACHE API minor version supported
+/// by Triton.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_ApiVersion(
+    uint32_t* major, uint32_t* minor);
+
+///
+/// Cache Lifetime Management
+///
+
+/// Create a new cache object. The caller takes ownership of the
+/// TRITONCACHE_Cache object and must call
+/// TRITONCACHE_CacheDelete to release the object.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache Returns the new cache object.
+/// \param config The config options to initialize the cache with.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheNew(
+    TRITONCACHE_Cache** cache, TRITONSERVER_Message* config);
+
+/// Delete a cache object.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache The cache object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheDelete(
+    TRITONCACHE_Cache* cache);
+
+///
+/// Cache Usage
+///
+
+/// Inserts entry into cache at specified key, unless key already exists.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache The object that is used to communicate with the cache
+///              implementation through a shared library.
+/// \param entry The entry to be inserted into the cache.
+/// \return a TRITONSERVER_Error indicating success or failure.
+///         Specific errors will be up the cache implementation, but general
+///         error best practices that should be followed are as follows:
+///         - TRITONSERVER_ERROR_INVALID_ARG
+///           - bad argument passed, nullptr, etc.
+///         - TRITONSERVER_ERROR_ALREADY_EXISTS
+///           - key already exists and will not be inserted again
+///         - TRITONSERVER_ERROR_INTERNAL
+///           - internal errors
+///         - nullptr
+///           - success
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheInsert(
+    TRITONCACHE_Cache* cache, const char* key, TRITONCACHE_CacheEntry* entry);
```

Comment:
I don't see why you need `CacheEntry`, you can have insert and lookup directly with respect to a list of `CacheItem`. Is this just to avoid `***` for lookup?

Reply:
Yes `CacheEntry` was to avoid `***` for lookup.

Reply:
How about having the first CacheItem represent the total number of responses (n) and the number of outputs in each response (m_1..m_n), then Triton will consider the first `m_1` buffers as outputs of response 1 and so on?

Reply:
> How about having the first CacheItem represent the total number of responses (n) and the number of outputs in each response (m_1..m_n), then Triton will consider the first m_1 buffers as outputs of response 1 and so on?

This may not work unless the cache insertion is done differently. If the insertion is triggered whenever a response is returned, in the case of decoupled model, the first insertion doesn't know how many responses will there be. Unless it holds and perform insertion at once when knowing the last response has been received (`RESPONSE_COMPLETE` flag), the "holding" part may not as simple as insertion on every response

Reply:
I see. For decoupled models, are we going to insert into the cache whenever a response is sent? In that case what happens if another request arrives with the same inputs before the insertion of previous request is finished? I think we probably need to the "holding" part regardless of the API we have for the cache.

Reply:
Along these lines of future decoupled support, do you think we should include some sort of insertion behavior to support this?
```
typedef enum TRITONCACHE_InsertBehavior {
  // default: fail on inserting a key that already exists
  TRITONCACHE_UNIQUE,
  // support decoupled: append items to existing key (response1 added previously, response2,3,... added later)
  TRITONCACHE_APPEND, 
  // overwrite existing key if it exists: maybe useful for some cases, not sure yet
  TRITONCACHE_REPLACE
};

TRITONCACHE_CacheInsert(..., TRITONCACHE_InsertBehavior behavior)
```

Reply:
I think it would be easier on the Cache API part if Triton can handle the "holding" part and insert all the responses at once. IMO, making cache responsible for the holding part is a bit unclean. @GuanLuo/Ryan do you think we can store all the responses belonging to a request in the Triton core, so that we don't have to add a new API to handle decoupled models in cache?

Reply:
I agree that Triton should handle the holding because of the question that @Tabrizian raised,
> In that case what happens if another request arrives with the same inputs before the insertion of previous request is finished?

In the decoupled case, the "cache entry" shouldn't be considered ready until the last response is processed. From the perspective of cache implementation, it makes sense that "insertion" is a sign that the cache entry is ready. Otherwise, it is extra requirement for cache to implement "holding" state, and the insert behavior above doesn't really address that, it is instead indicating that the content of cache entry could be mutable after insertion.

I am okay with insert behavior though, it makes sense for cache in general, but I can't think of when Triton may use other behaviors on top of my head. And if you do want to add it, should note that all behaviors are optional, the cache implementation should return error for behaviors that are not supported.

Reply:
What is the conclusion on this thread? Are we going to flatten the cache items into cache buffers?

Reply:
I believe the choices mentioned to remove a level of indirection are:
- pass list of CacheItems instead of one CacheEntry (I believe this will require a triple pointer in API, not necessarily a bad thing, just avoiding if possible)
- flatten list of buffers into one buffer inside CacheItem (I believe this will have a few downsides below): 
    - extra allocation and extra copies to form contiguous buffer of concatenated responses on triton side before sending to cache
    - local cache will still request allocated buffer from fixed-size boost managed memory, so the pre-allocated contiguous buffer from triton will be wasted
    - I believe storing a larger contiguous buffer (of concatenated buffers) will be likely be less efficient and have more fragmentation issues in boost managed buffer

Leaving the way it is now was to avoid the issues listed above. Let me know your thoughts